### PR TITLE
Enumerable method features

### PR DESCRIPTION
**Pull request by YuvrajM108**
In this pull request we:
- Implemented, tested, and debugged our own versions of Ruby's enumerable methods.
- Fixed any linter errors that were detected.

### CODE REVIEW

**Review by Dipeshtwis — COMMENTED** [review]
# Status: Invalid ❌

Please fix the issue mentioned below and then request for code review

## Reason

- [x] Linters are failing. Kindly fix this and submit it for another code review.

(Feel free to contact me on slack by the handle of  **deep** if you have any additional questions).
Happy Coding
[Dipesh Kumar](https://deeptwist.tk) | [Twitter](https://twitter.com/97deepeshkumar)

**Comment by YuvrajM108** [comment]
> # Status: Invalid ❌
> Please fix the issue mentioned below and then request for code review
> 
> ## Reason
> * [ ]  Linters are failing. Kindly fix this and submit it for another code review.
> 
> (Feel free to contact me on slack by the handle of **deep** if you have any additional questions).
> Happy Coding
> [Dipesh Kumar](https://deeptwist.tk) | [Twitter](https://twitter.com/97deepeshkumar)

Thank you for your feedback. I will definitely fix the errors however I am waiting for my coding partner to finish the other methods. I told him to complete it on Friday and he still has not pushed anything to this repository. Is it possible for me to complete this project solo instead of waiting on him?

**Comment by Dipeshtwis** [comment]
> 
> 
> > # Status: Invalid ❌
> > Please fix the issue mentioned below and then request for code review
> > ## Reason
> > 
> > * [ ]   Linters are failing. Kindly fix this and submit it for another code review.
> > 
> > (Feel free to contact me on slack by the handle of **deep** if you have any additional questions).
> > Happy Coding
> > [Dipesh Kumar](https://deeptwist.tk) | [Twitter](https://twitter.com/97deepeshkumar)
> 
> Thank you for your feedback. I will definitely fix the errors however I am waiting for my coding partner to finish the other methods. I told him to complete it on Friday and he still has not pushed anything to this repository. Is it possible for me to complete this project solo instead of waiting on him?

First of all, It's not a good way to work in a collaboration that you told your partner to finish one part and you'll finish the other. if you have any kind of issue with your partner, It's better to ask the student success team that my partner has inconsistency and I want to work solo. If you don't know how to reach student success, you can click on the **support** button from your dashboard. Hope this will help.

**Comment by YuvrajM108** [comment]
> > > # Status: Invalid ❌
> > > Please fix the issue mentioned below and then request for code review
> > > ## Reason
> > > 
> > > * [ ]    Linters are failing. Kindly fix this and submit it for another code review.
> > > 
> > > (Feel free to contact me on slack by the handle of **deep** if you have any additional questions).
> > > Happy Coding
> > > [Dipesh Kumar](https://deeptwist.tk) | [Twitter](https://twitter.com/97deepeshkumar)
> > 
> > 
> > Thank you for your feedback. I will definitely fix the errors however I am waiting for my coding partner to finish the other methods. I told him to complete it on Friday and he still has not pushed anything to this repository. Is it possible for me to complete this project solo instead of waiting on him?
> 
> First of all, It's not a good way to work in a collaboration that you told your partner to finish one part and you'll finish the other. if you have any kind of issue with your partner, It's better to ask the student success team that my partner has inconsistency and I want to work solo. If you don't know how to reach student success, you can click on the **support** button from your dashboard. Hope this will help.

I have already initiated the process with student success since Saturday and I have sent a follow-up message this morning. So far I have not received any response.

File: enumerable_methods.rb [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    return to_enum(:my_each) unless block_given?
+
+    arr = self
+    i = 0
+    while i < arr.length
+      yield(arr[i])
+      i += 1
+    end
+    self
+  end
```
Comment (Berabjesus) [inline enumerable_methods.rb]: - [x] returns the range itself after calling the given block once for each element in self
```
  range = Range.new(5,50)
  block = proc { |num| num > 100 }
  p range.my_each(&block) # error
  p range.each(&block) # (5..50)
```

- [x] when a block is given when self is a hash behaves like #each calls the given block once for each element in self

```
  my_each_output = ''
  block = proc { |num| my_each_output += num.to_s }
  { a: 1, b: 2, c: 3, d: 4, e: 5 }.each(&block)
  each_output = my_each_output.dup
  my_each_output = ''
  { a: 1, b: 2, c: 3, d: 4, e: 5 }.my_each(&block)

  p my_each_output # ""
  p each_output # "[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"
```

File: enumerable_methods.rb [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    arr = []
+    my_each { |value| arr << value if yield(value) }
+    arr
+  end
```
Comment (Berabjesus) [inline enumerable_methods.rb]: - [x] returns an array containing all elements of enum for which the given block returns a true value::range

```
  range = Range.new(5,50)
  block = proc { |num| num > 100 }
  p range.my_select(&block) # error
  p range.select(&block) # [ ]

```

File: enumerable_methods.rb [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      return true
+    elsif pmtr.nil?
+      my_each { |x| return false if nil_or_false?(x) }
+    elsif !pmtr.nil?
+      my_each { |x| return false unless class_or_regexp?(x, pmtr) }
+    else
+      my_each { |x| return false unless x == pmtr }
+    end
+    true
+  end
```
Comment (Berabjesus) [inline enumerable_methods.rb]: - [x] returns false if the block returns false or nil::range
- [x] returns false if the block ever returns a value false or nil::range

```
  range = Range.new(5,50)
  block = proc { |num| num > 100 }
  p range.my_all?(&block) # error

```

```
 p [1,2,3,4,].my_all?(Numeric) # error
  p [1,2,3,4,].all?(Numeric) # true

```

```
  p [3,3,3,3].my_all?(3) # error
  p [3,3,3,3].all?(3) # true
```

File: enumerable_methods.rb [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    return to_enum unless block_given? || proc
+
+    arr = []
+    if proc
+      my_each { |x| arr << proc.call(x) }
+    else
+      my_each { |x| arr << yield(x) }
+    end
+    arr
+  end
```
Comment (Berabjesus) [inline enumerable_methods.rb]: - [x] returns a new array with the results of running a given block once for every element in enum::range

```
range = Range.new(1,5)
  block = proc { |num| num > 100 }
  p range.my_map(&block) # error
  p range.map(&block) # [false, false, false, false, false]

```

File: enumerable_methods.rb [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    end
+    startval
+  end
+end
+
+# rubocop: enable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+
+def multipy_els(n_arr)
+  n_arr.my_inject(1, '*')
+end
```
Comment (Berabjesus) [inline enumerable_methods.rb]: - [x] The name of the method should be `multiply_els` not `multipy_els` you missed an `l`

**Review by Berabjesus — CHANGES_REQUESTED** [review]
# Status: CHANGES REQUIRED :warning: :construction: 
Hi @YuvrajM108 a very awesome start :+1. Great job implementing most of the ruby enum methods.

### However, you do have a few changes to carry out before we proceed so please take a look at the things you have to work on and their respective review comments below.

### my_each 

- [x] when a block is given when self is a range returns the range itself after calling the given block once for each element in self.
- [x]  when a block is given when self is a range behaves like #each calls the given block once for each element in self
- [x] when a block is given when self is a hash behaves like #each calls the given block once for each element in self

### my_each_with_index -- (use test cases from my_each below)
- [x] when a block is given when self is an array behaves like #each_with_index calls the given block once for each element in self
- [x]  when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- [x]  when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self
- [x] when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

### my_select 
- [x] returns an array containing all elements of enum for which the given block returns a true value::range

### my_all? 
- [x] returns false if the block returns false or nil::range
- [x] when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
- [x]  when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

### my_any? -- (use test cases from my_all? below)
- [x] returns false if the block ever returns a value false or nil::range
- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
- [x]  when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

### my_none?  -- (use test cases from my_all? below)

- [x] returns false if the block ever returns true for all elements::range

### my_map 

- [x] returns a new array with the results of running a given block once for every element in enum::range

 ### multiply_els 

- [x] accepts an array as an argument and multiplies all the elements of the array together using #my_inject

#### **If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help** 
**HAPPY CODING :computer:[Bereket Beshane](https://bereketababu.me)** | [LinkedIn](https://linkedin.com/in/bereketbeshane) | [GitHub](https://github.com/Berabjesus)

File: enumerable_methods.rb [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    arr = self if is_a?(Array)
+    arr = my_to_a if is_a?(Range) || is_a?(Hash)
+
+    i = 0
+    while i < arr.length
+      yield(arr[i])
+      i += 1
+    end
+    self
+  end
```
Comment (george-shammar) [inline enumerable_methods.rb:19]: - [ ] When a block is given, when self is a hash, it should behave like #each, it should call the given block once for each element in self.

File: enumerable_methods.rb [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    arr = self if is_a?(Array)
+    arr = my_to_a if is_a?(Range) || is_a?(Hash)
+
+    i = 0
+    while i < arr.length
+      yield(i)
+      i += 1
+    end
+    self
+  end
```
Comment (george-shammar) [inline enumerable_methods.rb:34]: - [ ] When a block is given, when self is an array, it should behave like #each_with_index, it should call the given block once for each element in self.
- [ ] When a block is given, when self is a range, it should behave like #each_with_index, it should call the given block once for each element in self.
- [ ] When a block is given, when self is a hash, it should behave like #each_with_index, it should call the given block once for each element in self.

**Review by george-shammar — CHANGES_REQUESTED** [review]
### **STATUS: CHANGES REQUIRED** :recycle: :recycle: 

You've done a really awesome job. Please keep it up. :+1: :+1: 

Don't be tired now. Make the few changes highlighted in my inline comments and you'll be good to go.

Most of the required changes are in your #my_each_with_index method.

When you're ready, please submit another CRR.

Please reach out to me on slack if you require further clarification or assistance.

Happy Coding. :+1:

**Review by george-shammar — CHANGES_REQUESTED** [review]
### **STATUS:CHANGES REQUIRED**  :recycle: :recycle: 

It's nice that you've implemented some of the changes from the previous review. :+1:  

Following our earlier chat on slack, I've dropped a message for you.

Please fix the remaining two edge cases below.

**my_each**
- [ ]  When a block is given, when self is a hash, it should behave like #each, it should call the given block once for each element in self. It seems the || operator you have on line 19 is not functional.
 Tip: Choose a simple solution over a complicated alternative.

**my_each_with_index**
- [ ] When a block is given, when self is a hash, it should behave like #each_with_index, it calls the given block once for each element in self.

Please submit another CRR once you're ready.

**Review by jubaan — CHANGES_REQUESTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> ⚠️ 🛠️ ⚠️ 🛠️ Changes Required 🛠️ ⚠️ 🛠️ ⚠️ </h1>
<br />
<p><b>"Fail fast, fail often"</b> and don't let this discourage you, because <b>only by failing do we learn to succeed</b>.</p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#congrats"> 💯 Awesome work 👏🏼 👏🏼 </a></li>
    <li><a href="#mandatory"> ⚠️ Things to Improve </a> 🚨 🚨 </li>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<h3 id="congrats"> 1. :100: Awesome work</h3>
<details close="close">
<summary> [click to open] Big or little achievements, but achievements at the end. <b>Feel proud that you're learning and improving everyday.</b></summary>



</details>

<hr>
<h3 id="mandatory"> 2. ⚠️ Things to Improve (or the To-Do list to get the project approved) </h3> 
<details open="open">
<summary> [click to open] Working on these observations will help you advance and get closer to the goal 🎯, postpone them or do them half-ways and they'll haunt you 👻 👻... FOREVER!! (It will feel like that, trust me). </summary>


- [ ] Remove all debugging code from enumerables file. 
- [ ]  enumerables `#my_each` when a block is given when self is a hash behaves like `#each` calls the given block once for each element in self
- [ ] enumerables `#my_each_with_index` when a block is given when self is a hash behaves like `#each_with_index` calls the given block once for each element in self



</details>
<p><b>NOTE</b>This list is or could be divided in what is show in this comment and inline comments posted below.<p>
<hr>

<h3 id="optional"> 3. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] </h3></summary>


- [x] Nothing more to add.


</details>

<h3 id="contact"> 4. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 5. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h2 align="center"> Keep up the great work :muscle: and Happy Coding :smiley: :computer:!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Review by mricanho — APPROVED** [review]
# Status: ✔️ Project Approved ✔️ 

Hello @YuvrajM108, awesome job on your rspecs! 🔥 

You worked hard in all the instructions that the previous Code Reviewers give you👍🏽
 
nice


Now you can merge this branch 🚀 

You can contact me on slack if you have any questions, just search for me as @ Miguel Ricaño.

### 💻 👩🏽‍💻  Happy Coding!  👨🏽‍💻 🖥️

**Pull request merged**